### PR TITLE
Check deferred nodes after all first passes in an import cycle

### LINKED ISSUE #481: Make it easier to work with circular module dependencies
It's sometimes painful to work around issues related to circular module dependencies. Some examples:

 * Mypy complains about not being able to infer a type because of a circular dependency.
 * Mypy type checks okay but at runtime something is undefined because of a circular dependency.
 * A type annotation introduces a circular module dependency where there was none before.

Not sure how these should be resolved. Some ideas:

 * Add a flag that dumps information about any circular module dependencies in the target program. This should help with debugging.
 * Warn about 'from imports' that are part of a circular module dependency, as these could fail at runtime.

### PR DESCRIPTION
**Pull request by gvanrossum**
Might fix #481.

### CODE REVIEW

File: test-data/unit/check-modules.test [hunk 1]
```diff
@@ -1175,6 +1175,33 @@ reveal_type(x)
...
+import a
+[file a.py]
+import b
+def f() -> int:
+    return b.x
+y = 0 + 0
+[file b.py]
+import a
+def g() -> int:
+    return a.y
```
Comment (JukkaL) [inline test-data/unit/check-modules.test]: Reveal the type of `a.y` (to verify that it doesn't somehow fall back to `Any`).

File: test-data/unit/check-modules.test [hunk 2]
```diff
@@ -1175,6 +1175,33 @@ reveal_type(x)
...
+x = 1 + 1
+
+[case testSymmetricImportCycle2]
+import b
+[file a.py]
+import b
+def f() -> int:
+    return b.x
```
Comment (JukkaL) [inline test-data/unit/check-modules.test]: Reveal the type of `b.x`.

File: mypy/checker.py [hunk 3]
```diff
@@ -121,55 +121,51 @@ class TypeChecker(NodeVisitor[Type]):
...
+        if options.strict_optional_whitelist is None:
+            self.suppress_none_errors = not options.show_none_errors
         else:
             self.suppress_none_errors = not any(fnmatch.fnmatch(path, pattern)
                                                 for pattern
-                                                in self.options.strict_optional_whitelist)
+                                                in options.strict_optional_whitelist)
+
+    def check_first_pass(self) -> None:
+        """Type check the entire file, but defer functions with forward references."""
```
Comment (JukkaL) [inline mypy/checker.py]: Be more explicit about forward reference mean. In particular, they mean references to definitions whose types haven't been inferred yet.

**Comment by JukkaL** [comment]
Looks good to me! This hopefully fixes one of the biggest pain points mypy users with large code bases have had.

I'd like to see some more test cases before merging:

1. Add test case where two passes aren't enough, to verify that we still
generate cannot determine type errors when needed. 

2. Add test case for referring to a decorated function whose type might not be inferred
(another common case where we had trouble before).

3. Add test case where we generate an error only on the second pass (to verify
that the import context is shown correctly). For example, something like this (partial example):
```
def f() -> None:
    a = b.x  # Type of b.x only available during second pass
    a + 1   # Error that we can't report during first pass
```

**Comment by gvanrossum** [comment]
I'll add more tests. I also discovered that reveal_type() needs to check for current_node_deferred.

**Comment by gvanrossum** [comment]
Rebased, generalized the multi-pass machinery, and added a test that would require 3 passes.

Still needs tests for (2) and (3). Also, there are still some print() calls left.

**Comment by gvanrossum** [comment]
I could use a hint for (2).

**Comment by gvanrossum** [comment]
I think I have all the required tests now. Jukka, can you double-check?

File: mypy/checker.py [hunk 4]
```diff
@@ -181,21 +187,33 @@ def visit_file(self, file_node: MypyFile, path: str, options: Options) -> None:
...
-        """Run second pass of type checking which goes through deferred nodes."""
-        self.pass_num = 1
-        for node, type_name in self.deferred_nodes:
+        This goes through deferred nodes, returning True if there were any.
+        """
+        if not self.deferred_nodes:
+            return False
+        self.errors.set_file(self.path)
+        self.pass_num += 1
+        print('---', self.path, 'pass', self.pass_num + 1, '---')
```
Comment (JukkaL) [inline mypy/checker.py]: Remove print.

File: mypy/checker.py [hunk 5]
```diff
@@ -181,21 +187,33 @@ def visit_file(self, file_node: MypyFile, path: str, options: Options) -> None:
...
+        todo = self.deferred_nodes
+        self.deferred_nodes = []
+        done = set()  # type: Set[FuncItem]
+        for node, type_name in todo:
+            if node in done:
+                continue
+            print(type_name, '.', node.fullname() or node.name())
```
Comment (JukkaL) [inline mypy/checker.py]: Remove another print.

File: test-data/unit/check-modules.test [hunk 6]
```diff
@@ -1186,6 +1186,89 @@ reveal_type(x)
...
+    def f2(self) -> None:
+        self.x2 = b.b
+[file b.py]
+import a
+b = 1 + 1
+[out]
+tmp/a.py:4: error: Cannot determine type of 'x2'
+
+[case testErrorInPassTwo]
+# flags: --hide-error-context
```
Comment (JukkaL) [inline test-data/unit/check-modules.test]: Don't hide error context to test the it works correctly for errors reported in the second pass.

File: test-data/unit/check-modules.test [hunk 7]
```diff
@@ -1186,6 +1186,89 @@ reveal_type(x)
...
+[file a.py]
+import b
+def f() -> None:
+    a = b.x + 1
+    a + ''
+[file b.py]
+import a
+x = 1 + 1
+[out]
+tmp/a.py:4: error: Unsupported operand types for + ("int" and "str")
```
Comment (JukkaL) [inline test-data/unit/check-modules.test]: Maybe also test this the other way as well (error reported in `b.py` in addition `a.py`.).

File: test-data/unit/check-modules.test [hunk 8]
```diff
@@ -1186,6 +1186,89 @@ reveal_type(x)
...
+
+[case testDeferredDecorator]
+# flags: --hide-error-context
+import a
+[file a.py]
+import b
+def g() -> None:
+    f()
```
Comment (JukkaL) [inline test-data/unit/check-modules.test]: Make sure that the decorator function returns an interesting type (not `Any`) and do `reveal_type(f())` here to verify that decorator got applied correctly.

**Comment by gvanrossum** [comment]
Ready for review/merge.

**Comment by JukkaL** [comment]
LGTM now -- thanks for the updates!

**Comment by gvanrossum** [comment]
OK, rebased and made all requested amendments.

**Pull request merged**